sparse_strips/vello_hybrid/src/scene.rs

```diff
@@ -368,3 +372,237 @@ impl GlyphRenderer for Scene {
+                }
+                RenderCommand::FillRect(rect) => {
+                    let path = rect.to_path(DEFAULT_TOLERANCE);
+                    self.generate_fill_strips(&path, &mut collected_strips, &mut collected_alphas);
+                    let count = collected_strips.len() - start_index;
+                    strip_ranges.push((start_index, count));
+                }
+                RenderCommand::StrokeRect(rect) => {
+                    let path = rect.to_path(DEFAULT_TOLERANCE);
+                    self.generate_stroke_strips(
+                        &path,
+                        &mut collected_strips,
+                        &mut collected_alphas,
+                    );
+                    let count = collected_strips.len() - start_index;
+                    strip_ranges.push((start_index, count));
+                }
+                RenderCommand::SetTransform(transform) => {
+                    self.transform = *transform;
+                }
+                RenderCommand::SetFillRule(fill_rule) => {
+                    self.fill_rule = *fill_rule;
+                }
+                RenderCommand::SetStroke(stroke) => {
+                    self.stroke = stroke.clone();
+                }
+                _ => {}
+            }
+        }
+
+        self.restore_state(saved_state);
+        (collected_strips, collected_alphas, strip_ranges)
+    }
+
+    fn execute_recording(&mut self, recording: &Recording) {
+        if let Some((cached_strips, cached_alphas)) = recording.get_cached_strips() {
+            let adjusted_strips = self.prepare_cached_strips(cached_strips, cached_alphas);
+
+            // Use pre-calculated strip ranges from when we generated the cache
+            let strip_ranges = recording.get_strip_ranges();
+            let mut range_index = 0;
+
+            // Replay commands in order, using cached strips for geometry
+            for command in &recording.commands {
+                match command {
+                    RenderCommand::FillPath(_)
+                    | RenderCommand::StrokePath(_)
+                    | RenderCommand::FillRect(_)
+                    | RenderCommand::StrokeRect(_) => {
+                        if range_index < strip_ranges.len() {
+                            let (start, count) = strip_ranges[range_index];
+
+                            if start < adjusted_strips.len() && count > 0 {
+                                let end = (start + count).min(adjusted_strips.len());
+                                let paint = self.encode_current_paint();
+                                self.wide.generate(
+                                    &adjusted_strips[start..end],
+                                    self.fill_rule,
+                                    paint,
+                                    0,
+                                );
+                            }
+
+                            range_index += 1;
```

Comment:
```suggestion
                        assert!(range_index < strip_ranges.len());
                        let (start, count) = strip_ranges[range_index];
                        assert!(start < adjusted_strips.len() && count > 0);

                        let end = start + count;
                        let paint = self.encode_current_paint();
                        self.wide
                            .generate(&adjusted_strips[start..end], self.fill_rule, paint, 0);

                        range_index += 1;
```

It's better to panic than to silently fail when an invariant is disobeyed. Rust has bounds checks (unless we do an unchecked lookup), so I think it's better to just `start + count` instead of do the minimum check (and potentially silently fail).